### ISSUE TITLE
(feature) Allow configurable token expiration leeway in auth token expiration check

### ISSUE THREAD

**Issue by amacleay-cohere** [guillp/requests_oauth2client#18]
- `requests_oauth2client` version: @ main
- Python version: 3.11.4
- Operating System: linux

### Description

Suppose that I have:
* A resource server with a system clock that is 0.1s fast
* An application client server with a system clock that is 0.1s slow
* A token held by the app client server that will expire the next clock tick

This token will be rejected by the resource server, even though the client thinks it has not yet expired.

Typically, clients implement a "clock skew" parameter that smoothly avoids expiration issues due to slight clock skew or use of tokens which are in the very act of expiring.

### What I Did

(This is a missing feature so just using demo code that might reproduce)

```python
import requests
from requests_oauth2client import OAuth2Client, OAuth2ClientCredentialsAuth

oauth2client = OAuth2Client(
    token_endpoint="https://url.to.the/token_endpoint",
    auth=("client_id", "client_secret"),
)

auth = OAuth2ClientCredentialsAuth(
    oauth2client, scope="myscope", resource="https://myapi.local"
)

# use it like this:
requests.get("https://myapi.local/resource", auth=auth)

# or like this:
session = requests.Session()
session.auth = auth

resp = session.get("https://myapi.local/resource")

# Wait exactly long enough for the token to expire in a couple nanoseconds
resp = session.get("https://myapi.local/resource") # Fails!
```

**Comment by guillp** [comment]
Thanks for the PR! That feature makes sense. I'll see how I can implement it soon.

**Comment by guillp** [comment]
Please review and test PR #20. This will allow setting the leeway (in seconds) in each Auth Handler, like this:

```python
auth = OAuth2ClientCredentialsAuth(
    oauth2client, scope="myscope", resource="https://myapi.local", leeway=30
)
```

**Comment by amacleay-cohere** [comment]
I noticed an issue with token expiration - opening a new issue to address.

Thanks for the work on that quick PR! I haven't tested yet, plan to later this week.

**Comment by guillp** [comment]
merged #20, closing
